File: quwoquan_service/services/assistant-service/internal/assistant/skill_subscription/application/outbox_relay.go
package application

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"

	"quwoquan_service/services/assistant-service/internal/assistant/skill_subscription/domain/model"
	"quwoquan_service/services/assistant-service/internal/assistant/skill_subscription/domain/ports"
)

const subscriptionOutboxClaimLease = 30 * time.Second

type OutboxEventPublisher interface {
	PublishSkillSubscription(context.Context, ports.OutboxEvent) error
}

type OutboxRelay struct {
	outbox    ports.TransactionalOutbox
	publisher OutboxEventPublisher
	ownerID   string
	now       func() time.Time

	healthMu           sync.RWMutex
	lastSuccessfulScan time.Time
	lastFailure        error
}

func NewOutboxRelay(
	outbox ports.TransactionalOutbox,
	publisher OutboxEventPublisher,
) (*OutboxRelay, error) {
	if outbox == nil || publisher == nil {
		return nil, errors.New("skill subscription outbox and publisher are required")
	}
	return &OutboxRelay{
		outbox: outbox, publisher: publisher,
		ownerID: "skill-subscription-relay-" + uuid.NewString(), now: time.Now,
	}, nil
}

func (relay *OutboxRelay) Drain(ctx context.Context, limit int) (int, error) {
	if limit <= 0 || limit > 200 {
		limit = 100
	}
	published := 0
	for published < limit {
		now := relay.now().UTC()
		event, found, err := relay.outbox.ClaimPendingOutbox(
			ctx, relay.ownerID, now, subscriptionOutboxClaimLease,
		)
		if err != nil {
			relay.recordFailure(err)
			return published, err
		}
		if !found {
			relay.recordSuccessfulScan(now, published > 0)
			return published, nil
		}
		if err := validateSubscriptionOutboxEvent(event); err != nil {
			failedAt := relay.now().UTC()
			retryErr := relay.outbox.ScheduleOutboxRetry(
				ctx, event.EventID, relay.ownerID,
				failedAt, failedAt.Add(outboxRetryDelay(event.AttemptCount)), "invalid_event",
			)
			if retryErr != nil {
				err = errors.Join(err, retryErr)
			}
			relay.recordFailure(err)
			return published, err
		}
		if err := relay.publisher.PublishSkillSubscription(ctx, event); err != nil {
			failedAt := relay.now().UTC()
			retryErr := relay.outbox.ScheduleOutboxRetry(
				ctx, event.EventID, relay.ownerID,
				failedAt, failedAt.Add(outboxRetryDelay(event.AttemptCount)), "publish_failed",
			)
			wrapped := fmt.Errorf("publish subscription event %s: %w", event.EventID, err)
			if retryErr != nil {
				wrapped = errors.Join(wrapped, retryErr)
			}
			relay.recordFailure(wrapped)
			return published, wrapped
		}
		if err := relay.outbox.MarkOutboxPublished(
			ctx, event.EventID, relay.ownerID, relay.now().UTC(),
		); err != nil {
			wrapped := fmt.Errorf("checkpoint subscription event %s: %w", event.EventID, err)
			relay.recordFailure(wrapped)
			return published, wrapped
		}
		published++
	}
	relay.recordSuccessfulScan(relay.now().UTC(), published > 0)
	return published, nil
}

func (relay *OutboxRelay) Run(ctx context.Context, interval time.Duration) {
	if interval <= 0 {
		interval = time.Second
	}
	relay.drainAndObserve(ctx)
	ticker := time.NewTicker(interval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			relay.drainAndObserve(ctx)
		}
	}
}

func (relay *OutboxRelay) Healthy(_ context.Context, maxStaleness time.Duration) error {
	if maxStaleness <= 0 {
		maxStaleness = 15 * time.Second
	}
	relay.healthMu.RLock()
	lastScan, lastFailure := relay.lastSuccessfulScan, relay.lastFailure
	relay.healthMu.RUnlock()
	if lastFailure != nil {
		return fmt.Errorf("skill subscription outbox relay unhealthy: %w", lastFailure)
	}
	if lastScan.IsZero() || relay.now().UTC().Sub(lastScan) > maxStaleness {
		return errors.New("skill subscription outbox relay heartbeat is stale")
	}
	return nil
}

func (relay *OutboxRelay) drainAndObserve(ctx context.Context) {
	if _, err := relay.Drain(ctx, 100); err != nil && ctx.Err() == nil {
		slog.ErrorContext(ctx, "skill subscription outbox drain failed", "err", err)
	}
}

func validateSubscriptionOutboxEvent(event ports.OutboxEvent) error {
	if err := validateExactSubscriptionPayloadKeys(event.Payload); err != nil {
		return err
	}
	var payload struct {
		SubscriptionID string `json:"subscriptionId"`
	}
	if err := json.Unmarshal(event.Payload, &payload); err != nil {
		return fmt.Errorf("decode subscription outbox payload: %w", err)
	}
	switch event.EventType {
	case model.EventCreated, model.EventStatusChanged, model.EventTriggered:
	default:
		return errors.New("unknown skill subscription event")
	}
	if strings.TrimSpace(event.EventID) == "" || strings.TrimSpace(event.AggregateID) == "" ||
		event.AggregateVersion <= 0 || event.OccurredAt.IsZero() ||
		strings.TrimSpace(payload.SubscriptionID) == "" || payload.SubscriptionID != event.AggregateID {
		return errors.New("skill subscription outbox event is incomplete")
	}
	return nil
}

func validateExactSubscriptionPayloadKeys(payload []byte) error {
	decoder := json.NewDecoder(bytes.NewReader(payload))
	token, err := decoder.Token()
	if err != nil || token != json.Delim('{') {
		return errors.New("skill subscription payload must be one JSON object")
	}
	seen := false
	for decoder.More() {
		keyToken, err := decoder.Token()
		key, ok := keyToken.(string)
		if err != nil || !ok || key != "subscriptionId" || seen {
			return errors.New("skill subscription payload keys do not match the event contract")
		}
		seen = true
		var value json.RawMessage
		if err := decoder.Decode(&value); err != nil {
			return fmt.Errorf("decode skill subscription id: %w", err)
		}
	}
	if _, err := decoder.Token(); err != nil || !seen {
		return errors.New("skill subscription payload omits subscriptionId")
	}
	var trailing json.RawMessage
	if err := decoder.Decode(&trailing); !errors.Is(err, io.EOF) {
		return errors.New("skill subscription payload has trailing content")
	}
	return nil
}

func outboxRetryDelay(attempt int) time.Duration {
	if attempt < 1 {
		attempt = 1
	}
	if attempt > 6 {
		attempt = 6
	}
	return time.Second * time.Duration(1<<(attempt-1))
}

func (relay *OutboxRelay) recordSuccessfulScan(at time.Time, recovered bool) {
	relay.healthMu.Lock()
	relay.lastSuccessfulScan = at
	if recovered {
		relay.lastFailure = nil
	}
	relay.healthMu.Unlock()
}

func (relay *OutboxRelay) recordFailure(err error) {
	relay.healthMu.Lock()
	relay.lastFailure = err
	relay.healthMu.Unlock()
}

File: quwoquan_service/services/recommendation-service/internal/recommendation/recommendation_model_release/application/outbox_relay.py
from __future__ import annotations

import json
import logging
import threading
from datetime import UTC, datetime, timedelta
from typing import Protocol
from uuid import uuid4

from ..domain.outbox import (
    OutboxEvent,
    TransactionalOutbox,
    validate_model_release_event_payload,
)

OUTBOX_LEASE_SECONDS = 30.0
_LOGGER = logging.getLogger(__name__)


class OutboxPublisher(Protocol):
    def publish(self, event: OutboxEvent) -> None: ...


class RecommendationModelReleaseOutboxRelay:
    def __init__(
        self,
        outbox: TransactionalOutbox,
        publisher: OutboxPublisher,
        *,
        interval_seconds: float = 1.0,
    ) -> None:
        if outbox is None or publisher is None:
            raise ValueError("model release outbox and publisher are required")
        if interval_seconds <= 0:
            raise ValueError("model release outbox interval must be positive")
        self._outbox = outbox
        self._publisher = publisher
        self._interval_seconds = interval_seconds
        self._owner_id = f"recommendation-model-release-relay-{uuid4()}"
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None
        self._health_lock = threading.Lock()
        self._last_successful_scan: datetime | None = None
        self._last_failure: Exception | None = None

    @staticmethod
    def _now() -> datetime:
        return datetime.now(UTC)

    def drain(self, limit: int = 100) -> int:
        bounded_limit = limit if 0 < limit <= 200 else 100
        published = 0
        while published < bounded_limit:
            now = self._now()
            event = self._outbox.claim_pending_outbox(
                self._owner_id,
                now,
                OUTBOX_LEASE_SECONDS,
            )
            if event is None:
                self._record_success(now, recovered=published > 0)
                return published
            try:
                self._validate(event)
            except Exception as error:
                self._schedule_retry(event, now, "invalid_event")
                self._record_failure(error)
                raise
            try:
                self._publisher.publish(event)
            except Exception as error:
                self._schedule_retry(event, now, "publish_failed")
                self._record_failure(error)
                raise
            try:
                self._outbox.mark_outbox_published(
                    event.event_id,
                    self._owner_id,
                    self._now(),
                )
            except Exception as error:
                self._record_failure(error)
                raise
            published += 1
        self._record_success(self._now(), recovered=published > 0)
        return published

    def _schedule_retry(
        self,
        event: OutboxEvent,
        now: datetime,
        failure_code: str,
    ) -> None:
        attempt = max(1, min(event.attempt_count, 6))
        self._outbox.schedule_outbox_retry(
            event.event_id,
            self._owner_id,
            now + timedelta(seconds=2 ** (attempt - 1)),
            failure_code,
        )

    @staticmethod
    def _validate(event: OutboxEvent) -> None:
        if (
            not event.event_id.strip()
            or not event.aggregate_id.strip()
            or event.aggregate_version <= 0
            or event.occurred_at is None
            or event.occurred_at.tzinfo is None
            or not isinstance(event.payload, dict)
            or not event.payload
        ):
            raise ValueError("model release outbox event is incomplete")
        validate_model_release_event_payload(event.event_type, event.payload)
        if event.payload["id"] != event.aggregate_id:
            raise ValueError("model release event aggregate identity does not match")
        json.dumps(
            event.payload,
            allow_nan=False,
            ensure_ascii=False,
            separators=(",", ":"),
            sort_keys=True,
        )

    def run_once(self) -> None:
        try:
            self.drain()
        except Exception:
            _LOGGER.exception("recommendation model release outbox drain failed")

    def _run(self) -> None:
        self.run_once()
        while not self._stop.wait(self._interval_seconds):
            self.run_once()

    def start(self) -> None:
        if self._thread is not None and self._thread.is_alive():
            return
        self._stop.clear()
        self._thread = threading.Thread(
            target=self._run,
            name="recommendation-model-release-outbox",
            daemon=True,
        )
        self._thread.start()

    def stop(self) -> None:
        self._stop.set()
        if self._thread is not None:
            self._thread.join(timeout=max(2.0, self._interval_seconds * 2))
        self._thread = None

    def healthy(self, *, max_staleness_seconds: float = 10.0) -> bool:
        with self._health_lock:
            last_scan = self._last_successful_scan
            last_failure = self._last_failure
        if last_failure is not None or last_scan is None:
            return False
        return (self._now() - last_scan).total_seconds() <= max_staleness_seconds

    def _record_success(self, at: datetime, *, recovered: bool) -> None:
        with self._health_lock:
            self._last_successful_scan = at
            if recovered:
                self._last_failure = None

    def _record_failure(self, error: Exception) -> None:
        with self._health_lock:
            self._last_failure = error


__all__ = ["RecommendationModelReleaseOutboxRelay"]
